### ISSUE TITLE
add emailAdapter to ENV variables

### ISSUE THREAD

**Issue by Cliffordwh** [ParsePlatform/parse-server#1277]
Is it possible to add the `emailAdapter` to the env variables?  like we have done with push and files etc.

I'm not making use of the `index.js `file from the sample. Firing off a json config.  So not sure where to add the mailgun adapter with its vars.

Thanks

**Comment by drew-gross** [comment]
If you are using a JSON config env var, you can configure your email adapter like this:

```json
  "emailAdapter": {
    "module": "parse-server-simple-mailgun-adapter",
    "options": {
      "fromAddress": "parse@example.com",
      "domain": "example.com",
      "apiKey": "key-mykey"
    }
  }
```

**Comment by Cliffordwh** [comment]
@drew-gross  Set it as described but the logs for parse-server show the following when i try reset:

`Uncaught internal server error. Trying to send a reset password but no adapter is set undefined
`

My json looks as follows:
```
{
  "apps" : [{
    "name"        : "parse-server",
    "script"      : "/usr/bin/parse-server",
    "watch"       : true,
    "merge_logs"  : true,
    "cwd"         : "/home/parse",
    "env": {
      "PARSE_SERVER_CLOUD_CODE_MAIN": "/home/parse/cloud/main.js",
      "PARSE_SERVER_DATABASE_URI": "xxxx",
      "PARSE_SERVER_APPLICATION_ID": "xxx",
      "PARSE_SERVER_MASTER_KEY": "xx",
      "PARSE_SERVER_FILE_KEY": "xxxx",
      "PARSE_SERVER_JAVASCRIPT_KEY": "xxxx",
      "PARSE_SERVER_REST_API_KEY": "xxxx",
      "PARSE_SERVER_MOUNT_PATH" : "",
      "PARSE_SERVER_FILES_ADAPTER" : "parse-server/lib/Adapters/Files/S3Adapter",
      "S3_ACCESS_KEY" : "xxx",
      "S3_SECRET_KEY"  : "xxxxxx",
      "S3_BUCKET" : "xxxxx",
      "S3_DIRECT_ACCESS" : "true",
      "PARSE_SERVER_FACEBOOK_APP_IDS" : "xxxxxxx",
      "PARSE_SERVER_PUSH" :  {"adapter" : "./Push/OneSignalPushAdapter",
                                             "oneSignalAppId": "xxxxxxx",
                                             "oneSignalApiKey": "xxxxxxxxx"
                             },
      "emailAdapter" : {
           "module": "parse-server-simple-mailgun-adapter",
                  "options": {
                  "fromAddress": "xxxxxx",
                  "domain": "xxxxxx",
                  "apiKey": "xxxxxxx"
    }
      }
    }
  }]
}
```

**Comment by flovilmart** [comment]
did you add `parse-server-simple-mailgun-adapter` to your package?

**Comment by Cliffordwh** [comment]
@flovilmart how would i do that? I did run  `npm i parse-server-simple-mailgun-adapter`

**Comment by flovilmart** [comment]
ok you don't have a base package... how did you setup parse-server?

**Comment by Cliffordwh** [comment]
i used the global install    npm - g !

**Comment by flovilmart** [comment]
so install with -g :)

**Comment by Cliffordwh** [comment]
i did, thanks. But still get the ```
"Uncaught internal server error. Trying to send a reset password but no adapter is set undefined
"
```

**Comment by Cliffordwh** [comment]
@flovilmart is there a reason that emailAdapter is not listed in the cli-definitions file?

**Comment by flovilmart** [comment]
maybe because it's missing :)

**Comment by Cliffordwh** [comment]
lol, could that be the reason why its not available?

**Comment by flovilmart** [comment]
I'll post a fix :P

**Comment by Cliffordwh** [comment]
Thanks

**Comment by Cliffordwh** [comment]
deff solved my problem, all working now! . Now i can just make use of `PARSE_SERVER_EMAIL_ADAPTER`. or whatever you name it. 

```
"emailAdapter": {
    env: "PARSE_SERVER_EMAIL_ADAPTER",
    action: function action(opt) {
      return JSON.parse(opt);
    }
  },
```

**Comment by Cliffordwh** [comment]
Thanks again. will wait on the update!

**Comment by flovilmart** [comment]
Does it work with the emailAdapter option in --emailAdapter or config?

**Comment by Cliffordwh** [comment]
works with option. As above.

**Comment by junal** [comment]
I'm having exact same problem. I migrated an app to digitalocean following this tuturial https://www.digitalocean.com/community/tutorials/how-to-migrate-a-parse-app-to-parse-server-on-ubuntu-14-04

Can you please tell me where to add emailAdapter.

**Comment by Cliffordwh** [comment]
@junal  The changes to the CLI-Definitions will be available in the next update. This will allow you then to set your eco file with the settings above.

**Comment by junal** [comment]
@Cliffordwh Thanks! I have updated version of parse-server. So I added the ENV variable like this at /home/parse/ecosystem.json file

`{
  "apps": [
    {
      "name": "parse-wrapper",
      "script": "/usr/bin/parse-server",
      "watch": true,
      "merge_logs": true,
      "cwd": "/home/parse",
      "env": {
        "PARSE_SERVER_CLOUD_CODE_MAIN": "/home/parse/cloud/main.js",
        "PARSE_SERVER_DATABASE_URI": "XXX",
        "PARSE_SERVER_APPLICATION_ID": "XXX",
        "PARSE_SERVER_MASTER_KEY": "XXX",
        "PARSE_SERVER_EMAIL_ADAPTER": "parse-server-simple-mailgun-adapter"
      }
    }
  ]
}
`
But i still get the error message..."error: Uncaught internal server error. Trying to send a reset password but no adapter is set undefined".

What am I missing?

**Comment by flovilmart** [comment]
Those changes have't been release yet on NPM. They will be part of the next update

**Comment by junal** [comment]
@flovilmart I have migrated an app from parse.com to Digitalocean server following their tutorial. Everything is working fine except few params are not working in the reset password and I'm not getting the verification email. Here's how my ecosystem.json looks like...

```
{
  "apps": [
    {
      "name": "parse-wrapper",
      "script": "/usr/bin/parse-server",
      "watch": true,
      "merge_logs": true,
      "cwd": "/home/parse",
      "verifyUserEmails": true,
      "publicServerURL": "http://example.com:1337/parse",
      "appName": “appName”,
      "env": {
        "PARSE_SERVER_CLOUD_CODE_MAIN": "/home/parse/cloud/main.js",
        "PARSE_SERVER_DATABASE_URI": “XXX”,
        "PARSE_SERVER_APPLICATION_ID": "XXX",
        "PARSE_SERVER_MASTER_KEY": "XXX",
        "PARSE_SERVER_EMAIL_ADAPTER": {
          "module": "parse-server-simple-mailgun-adapter",
          "options": {
            "fromAddress": "no-reply@example.com”,
            "domain": “example.com”,
            "apiKey": "XXX"
          }
        },
        "PARSE_SERVER_FILE_KEY": "XXX",
      }
    }
  ]
}

```

And here's how I'm getting email https://dl.dropboxusercontent.com/spa/e8jedo3lnl7g7wc/rcu187al.png it seems appName, publicServerURL and verifyUserEmails are not working. Do you know where I'm making mistake?

Thanks

**Comment by flovilmart** [comment]
You should define those properties from their env counterpart.

**Comment by Cliffordwh** [comment]
@junal Also ensure you are running 2.2.5. 

If you need to make use of appName urgently as i did, you can add it to the CLI-Definitions yourself as a env or wait for the next update.

**Comment by junal** [comment]
@Cliffordwh I updated to 2.2.5 and added appName in the cli-definitions as env i.e APP_NAME but no luck yet. Email still comes like this https://dl.dropboxusercontent.com/spa/e8jedo3lnl7g7wc/-7qtpyq1.png and where to add verifyUserEmails for verification email.

**Comment by Cliffordwh** [comment]
@junal this because your json file is incorrect based on the sample you posted above. Can you share your latest file?

**Comment by junal** [comment]
@Cliffordwh 

```
{
  "apps": [
    {
      "name": "parse-wrapper",
      "script": "/usr/bin/parse-server",
      "watch": true,
      "merge_logs": true,
      "cwd": "/home/parse",
      "env": {
        "PARSE_SERVER_CLOUD_CODE_MAIN": "/home/parse/cloud/main.js",
        "PARSE_SERVER_DATABASE_URI": “XXX”,
        "PARSE_SERVER_APPLICATION_ID": “XX”,
        "PARSE_SERVER_MASTER_KEY": “XX”,
        "PARSE_SERVER_EMAIL_ADAPTER": {
          "module": "parse-server-simple-mailgun-adapter",
          "options": {
            "fromAddress": "no-reply@example.com”,
            "domain": "sandboxXXX.mailgun.org",
            "apiKey": "key-XXX”
          }
        },
        "PARSE_SERVER_URL": "http://example.com:1337/parse",
        "PARSE_PUBLIC_SERVER_URL": "http://example.com:1337/parse",
        "APP_NAME": “App Name”,
        "PARSE_SERVER_FILE_KEY": “XXX”
      }
    }
  ]
}
```

**Comment by Cliffordwh** [comment]
Are you using PM2? You need to then first stop the process, pm2 kill , pm2 dump then restart. Then save pm2 save. Just to ensure the latest json is added.

See if that solves it.

**Comment by flovilmart** [comment]
also APP_NAME should be PARSE_SERVER_APP_NAME

**Comment by junal** [comment]
@Cliffordwh @flovilmart Thank you so much!!! All working now.

**Comment by junal** [comment]
@Cliffordwh @flovilmart Suddenly, reset password and verification emails aren't working. It says "error: Uncaught internal server error. Trying to send a reset password but no adapter is set undefined"

I didn't make any changes in the ENV. What could cause this issue?

**Comment by drew-gross** [comment]
@junal we have lots of test cases that cover reset emails. Can you open a new issue with your configuration and a snippet of code that demonstrates the error?

**Comment by majidhassan** [comment]
I tried the configuration posted by @junal but I keep getting `SimpleMailgunAdapter requires an API Key, domain, and fromAddress`. I'm using parse-server 2.2.16. Any suggestions?

**Comment by tanzeelrana** [comment]
`parse-server-simple-mailgun-adapter: command not found`

**Comment by supermar1010** [comment]
I just struggled with setting an mail adapter via an env variable for about 2 hours. I'm setting the env variables in a docker-compose. The important thing was to use double quotes for attributes. 

This is how it works:

```version: '3'
services:
  parse:
    image: parseplatform/parse-server
    environment:
      ....
      - PARSE_SERVER_VERIFY_USER_EMAILS=true
      - PARSE_SERVER_PREVENT_LOGIN_WITH_UNVERIFIED_EMAIL=true
      - 'PARSE_SERVER_EMAIL_ADAPTER={"module": "parse-server-generic-email-adapter", "options": {"service": "Gmail", "email": "my@email.com", "password": "myPassword"}}'
```

The module needs to be installed before, I did this via a custom dockerfile as follows:

```
FROM parseplatform/parse-server

USER root
RUN npm install --save parse-server-generic-email-adapter
USER node

```
